### DISCUSSION TITLE
Feature Request: HTTPFS - Document advanced usage of Glob

### DISCUSSION THREAD

**Discussion by dberardo-com** [duckdb/duckdb#5590] [Ideas]
I am trying to achieve time-range based scan of parquet / csv files using globs in time/hive-style partitioned S3 buckets. Due to the limited support of time ranges in regex i cannot find a robust way to query time ranges.

I guess that adequate documentation would be great for other people struggling.

Here is an example of what i am trying to achieve:

```
# Aim: try to query only partitions from Dec (12/2022) to Jul 2023 (09/2023)

Glob regex: year=[2-2][0-0][2-2][2-4](.*)month=[0-9][1-2]

asd/year=2022/month=01/day=02/ererer.parquet
asd/year=2024/month=01/day=02/ererer.parquet
asd/year=2025/month=01/day=02/ererer.parquet
asd/year=2024/month=01/day=02/ererer.parquet
asd/year=2024/month=10/day=02/ererer.parquet
asd/year=2024/month=12/day=02/ererer.parquet

```

**Comment by dberardo-com** [comment]
i hope this is enough to understand what i am trying to achieve here, let me know if more details are needed

**Comment by Mytherin** [comment]
You are probably looking for the [hive partitioning flag](https://duckdb.org/docs/extensions/httpfs.html#hive-partitioning):

```sql
SELECT * FROM parquet_scan('asd/*/*/*/*.parquet', HIVE_PARTITIONING = 1);
```

This allows you to query the columns normally instead of having to use a regex on the file names.

**Reply by dberardo-com** [reply]
i am sorry if my description was misleading, but i am already able to read in hive partitioned files. what i am trying to achieve is a sort of pre-filtering of files to read using globs.

in my case the S3 buckets are hive-partitioned by time, as in:

`bucket/year=2022/month=12/day=04/file.parquet`

and now i would like to reduce my glob selection to only those path that a fall between a specific interval. foe example if i have these 3 files:

```
bucket/year=2022/month=12/day=04/file.parquet
bucket/year=2022/month=12/day=05/file.parquet
bucket/year=2022/month=12/day=06/file.parquet
```

and i want only to select the ones between 04/12/2022 and 05/12/2022, i would do this:

`parquet_scan("bucket/year=2022/month=12/day=0[4-5]/file.parquet", HIVE_PARTITIONING = 1);`

but this would not work well in case of month/year changing dates.

**Reply by Mytherin** [reply]
When using the `HIVE_PARTITIONING` flag adding filters on the hive-partitioned columns should automatically result in files being pruned. e.g. you should be able to do it like this:

```sql
SELECT *
FROM parquet_scan('asd/*/*/*/*.parquet', HIVE_PARTITIONING = 1)
WHERE make_date(year, month, day) >= DATE '2022-12-01' AND make_date(year, month, day) <= DATE '2023-06-01'
```

And the system will automatically figure out which files to read.

Is that not working for you?

**Reply by dberardo-com** [reply]
sure, i can filter by time using a where clause, i just thought that by doing it already in the Glob might have improved performance, but if you tell me it makes no difference, then i will apply the standard WHERE strategy. I just dont know much about the internals of DuckDB to judge that, so i thought i would simply give it a try and compare both ways.

Also, i was thinking to apply this strategy perhaps to other file types, like CSVs, where filtering on time would be expensive. But i guess there is no csv_scan function at the moment, so my attempt is just too "forward looking", isnt it ?

**Reply by Mytherin** [reply]
There is a CSV read function: https://duckdb.org/docs/guides/import/csv_import

**Reply by dberardo-com** [reply]
well sure, but this strategy would only work for reading multiple files at one, especially in the case of HIVE-partitioned scans. I hope I could make my concern clear.

**Reply by Mytherin** [reply]
The same syntax should work with CSV files.

**Comment by dberardo-com** [comment]
i have give a try to that where statement, and i think it does the job.

I am now facing a different issue, related to type casting and it seems i cannot get rid of it.

the problem is that i am using a JSON schema that declares a field "T" with type null OR int64 (this is a timestamp in unix epoch). However, sometimes wrong values are stored in the parquet files, basically empty strings T=''

and now if i try to select * from parquet_scan() i get this error:

```
[Error: Conversion Error: Could not convert string '' to INT64] {
  errno: -1,
  code: 'DUCKDB_NODEJS_ERROR'
}
```

I have tried using TRY_CAST both in SELECT and in WHERE but it seems the error is not going away. Also if i use WHERE T <> '' in the query, then the result is always empty ... although it should not be.

any workaround here? can i impose a different schema to a parquet_scan query to allow for string values ?

**Comment by dberardo-com** [comment]
alright, so i have given this another try, and end up with a strange behavior of the WHERE clause, have a look here: https://github.com/duckdb/duckdb/issues/5727
